File: clients/typescript/src/model/debug-state.ts
import { Stream } from "../stream";

/**
 * TODO - Document
 */
export class DebugState {
    /**
     * TODO - Document
     */
    pressedKeys: Array<string>

    constructor(pressedKeys: Array<string>) {
        this.pressedKeys = pressedKeys;
    }

    /**
     * Read DebugState from input stream
     */
    static async readFrom(stream: Stream): Promise<DebugState> {
        let pressedKeys;
        pressedKeys = [];
        for (let pressedKeysCount = await stream.readInt(); pressedKeysCount > 0; pressedKeysCount--) {
            let pressedKeysElement;
            pressedKeysElement = await stream.readString();
            pressedKeys.push(pressedKeysElement);
        }
        return new DebugState(pressedKeys)
    }

    /**
     * Write DebugState to output stream
     */
    async writeTo(stream: Stream) {
        let pressedKeys = this.pressedKeys;
        await stream.writeInt(pressedKeys.length);
        for (let pressedKeysElement of pressedKeys) {
            await stream.writeString(pressedKeysElement);
        }
    }
}

File: clients/ruby/model/debug_state.rb
module Model

# TODO - Document
class DebugState
    # TODO - Document
    attr_accessor :pressed_keys

    def initialize(pressed_keys)
        @pressed_keys = pressed_keys
    end

    # Read DebugState from input stream
    def self.read_from(stream)
        pressed_keys = []
        stream.read_int().times do |_|
            pressed_keys_element = stream.read_string()
            pressed_keys.push(pressed_keys_element)
        end
        DebugState.new(pressed_keys)
    end

    # Write DebugState to output stream
    def write_to(stream)
        stream.write_int(@pressed_keys.length())
        @pressed_keys.each do |pressed_keys_element|
            stream.write_string(pressed_keys_element)
        end
    end

    def to_s
        string_result = "DebugState { "
        string_result += "pressed_keys: "
        string_result += "[ "
        pressed_keys_index = 0
        @pressed_keys.each do |pressed_keys_element|
            if pressed_keys_index != 0
                string_result += ", "
            end
            string_result += pressed_keys_element.dump
            pressed_keys_index += 1
        end
        string_result += " ]"
        string_result += " }"
        string_result
    end

    def to_str
        to_s
    end
end

end